Page sizes and rotation

mvantzet opened the pull request:
This is a fix related to page sizes / rotation / coordinate transformations (issue #560):

The initial transformation matrix was incorrect, as it translated by the cropbox width/height instead of by the cropbox left/bottom offsets. Also, it did not translate the results back into the 1st quadrant so that (0,0) would (again) be the lower left corner origin for the cropped area. The result was that text positions did not match up with generated page images for pages that are cropped and rotated.

Also fixed an issue parsing page sizes (A3, A4, etc.) for (a) non-integer page sides and (b) landscape width/height. Added a small unit test to verify correct working.

The implementation of UserSpaceUnit is still left "todo". I think that it may be a simple matter of adding a scaling transformation to the initial matrix but I'm not 100% sure of the impact. (From what I understood user space units are not widely used.)

BobLd commented:
Hi @mvantzet thanks for the work. I've added some comments. Could you also add an integration test using the 'SPARC - v9 Architecture Manual.pdf' document? It's already in the integration tests document folder. I know the 1st page dimension where not correct before your fix.

I will try to find documents with UserSpaceUnit != 1 if you want to have a look too

BobLd commented:
[MOZILLA-11308-0.pdf](https://github.com/UglyToad/PdfPig/files/10948252/MOZILLA-11308-0.pdf)

If you wnat to have a look at inplementing UserSpaceUnit, have a look at the doc above

BobLd commented:
Hi @mvantzet, agreed for the 'separate issue / PR for user space units.'

I've made comments in your code, you can't see them?
Below is a screenshot
![image](https://user-images.githubusercontent.com/38405645/224848508-d69731c6-063f-4c96-bc4b-676ae9f098eb.png)

I will do some test with the annotations when I have time - do you have an example document with annotations that are off?

mvantzet commented on src/UglyToad.PdfPig/Annotations/AnnotationProvider.cs:
```diff
@@ -37,11 +40,12 @@ public IEnumerable<Annotation> GetAnnotations()
 
                 var contents = GetNamedString(NameToken.Contents, annotationDictionary);
                 var name = GetNamedString(NameToken.Nm, annotationDictionary);
-                var modifiedDate = GetNamedString(NameToken.M, annotationDictionary);
```
I will have to revert this change, reading up on the modified date in the PDF reference 8.4.1 which indicates this field can basically contain anything.

mvantzet commented:
Weird, I cannot see your comments. I added a comment myself just now, about reverting the change to the modified date... Can you see that?

Regarding your question about annotations that are off: I added a file in the tests with an annotation at the top-left, after rotation at the top-right, of the page.

BobLd commented:
@mvantzet thanks for adding the documents. I will have a look

Yes I can see your comment, very strange you can't see mine.
I've put 4 comments:
![image](https://user-images.githubusercontent.com/38405645/225405141-9fb36c27-ade8-44dd-a01a-3a60f843cf1d.png)
![image](https://user-images.githubusercontent.com/38405645/225405385-ff4bc8d4-c6fc-4217-9552-17420a8f7891.png)
![image](https://user-images.githubusercontent.com/38405645/225405532-f73f5283-601d-46ca-a3f1-fe7f280aaa63.png)

Plus the one you already saw and fixed

EliotJones reviewed (approved):
I'll trust your maths is correct here, despite doing maths for years vectors and matrices is something I never was able to grasp.

Not sure if you wanted to push a revert to the change on modified date in `AnnotationProvider`? Also GitHub checks is complaining about the expected/actual arguments to the asserts in `RotationAndCroppingTests` so if you're going to make any changes please swap those 🙏

EliotJones commented on src/UglyToad.PdfPig.Tests/Geometry/PdfRectangleTests.cs:
```diff
@@ -1,9 +1,11 @@
     using PdfPig.Core;
     using Xunit;
     using System.Collections.Generic;
+    using System.Drawing;
```
I didn't notice this was in tests

mvantzet commented:
I swapped the asset args. Regarding the "date modified": that was reverted in commit [048caed](https://github.com/UglyToad/PdfPig/pull/561/commits/048caedd649942ba357cfff14cce5e0b89cd9de6) above
